design/bgp-bfd.md

```diff
@@ -0,0 +1,91 @@
+# BFD enablement in BGP
+
+## Summary
+
+The purpose of this enhancement is to use enable [BFD](https://en.wikipedia.org/wiki/Bidirectional_Forwarding_Detection) to be used
+in conjunction with BGP.
+
+## Motivation
+
+BFD provides a quicker path failure detection than BGP, and using them together will allow users to provide a better
+service.
+
+BFD for BGP is supported by FRR [out of the box](http://docs.frrouting.org/en/latest/bfd.html#bgp-bfd-configuration).
+
+
+### Goals
+
+* Provide a way to establish BFD on BGP session.
+* Provide a way to configure the parameters of the BFD session associated with a BGP session.
+* Provide BFD on BGP only on top of the FRR implementation.
+
+### Non-Goals
+
+* Implement BFD on the legacy BGP implementation.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I want to declare a BGP session to be backed up by BFD.
+
+#### Story 2
+
+As a cluster administrator, I want to be able to set the BFD parameters related to a BGP session.
+
+## Design Details
+
+### BFD Design proposal
+
+The idea is to leverage FRR to enable BFD on a BGP session.
+
+When declaring a BFD peer, all it takes to enable BFD is to add a `bfd` property on the neighbour:
+
+```yaml
+    neighbor <A.B.C.D|X:X::X:X|WORD> bfd profile BFDPROF
+```
+
+The proposal here is to define a bfd profile section in the config structure that looks like:
+
+```yaml
+    bfd-profiles:
+    - name: bfdprofile1
+      receive-interval: 150
+      transmit-interval: 150
+      detect-multiplier: 10
+```
+
+When setting a BGP peer, an optional bfd-profile property will enable BFD:
+
+```yaml
+    peers:
+    - peer-address: 10.0.0.1
+      peer-asn: 64501
+      my-asn: 64500
+      bfd-profile: bfdprofile1
+```
+
```

Comment:
Would we need to communicate anything back to the cluster administrator. For example in the operator CRD status?

Reply:
This is a good point, and it really depends on where the CRD is implemented. It would be cool to be able to report the status of the session (and this is probably something we can get also for a BGP session?), but if the CRD is implemented in the operator we can't have the operator retrieve that information out of the speakers / frr container.

Reply:
Yeah this is a general problem that we will need to address.

Reply:
BGP peer can only select one BFD profile correct i.e 1:1 relationship ? if that is the case why don't we embed the BFD config to BGP-PEER object instead of creating new CR , we had similar to debate with @russellb about bgp-community and the decision was to avoid CR overhead , so just curious why this is any different ?

Reply:
BFD profile can fall in the category where it needs to be defined only in one peer like other global configuration routerId, myASN, port ......etc

Reply:
That is less likely to happen (at least compared to port, routerid, etc which are supposed to be unique).
So it's likely to be the same for multiple sessions. On top of that, that's the way FRR (which has more miles than us) handles that.
The only difference is that they leverage a default profile too iirc, but if you want to change the parameters you need to rely on a profile. Drifting from this will cause us a bit of a headache I am afraid.

Reply:
I'm OK with keeping this one separate.  The BGP communities config was a bit of a different case since that was such a trivial piece of config info (simple aliases).  The BFD config is a bit more involved and the common case is probably just a single profile.